docs/2024/puzzles/day09.md

```diff
@@ -1,15 +1,320 @@
+- `Some(index)` for file blocks, with their corresponding index
+- `None` for free blocks
+
+Our main driver converts the input to a `Disk`, create a new compact `Disk` from it and finally computes its checksum
+
+```scala
+def part1(input: String): Long =
+
+  type Disk = Seq[Option[Int]]
+  extension(disk: Disk)
+    def checksum: Long = ???
+
+  def createDisk(input: String): Disk = ???
+
+  def compact(disk: Disk): Disk = ???
+
+  val disk = createDisk(input)
+  compact(disk).checksum
+```
+
+Let's first implement `checksum`: It is the sum of each file ID times its position in the disk
+```scala
+extension(disk: Disk)
+  def checksum: Long = disk
+    .zipWithIndex
+    .map(_.getOrElse(0).toLong * _) // Free blocks are mapped to 0
+    .sum
+```
+
+To create our disk from the input, we need to:
+- Convert our input to a `List[Int]` ranging from 0 to 9
+- Group elements 2 by 2 to create pairs of (file, free) blocks count
+- Zip these groups with their indices
+- Unroll these pairs into an actual sequence with the correct number of elements
+- Concatenate these newly created sequences
+```scala
+def createDisk(input: String): Disk =
+    val intInput = input.toList.map(_ - '0') // Convert characters to int [0, 9]
+    val fileFreeGroups = intInput.grouped(2).toVector // Last group will contain a single element
+    val zippedGroups = fileFreeGroups.zipWithIndex
+    val disk = zippedGroups.flatMap:
+      case (List(fileN, freeN), i) => 
+        // File block followed by free block
+        List.fill(fileN)(Some(i)) ::: List.fill(freeN)(None)
+      case (List(fileN), i) => 
+        // Final file block
+        List.fill(fileN)(Some(i))
+      case _ => Nil
+    return disk
+```
+
+Finally, we need to compact the disk we obtain: Iterate over the disk elements, from the beginning (left)
+- If we encounter a free block: replace it with the last element in the disk and repeat the recursion
+- If we encounter a file block: Append it to the result and continue with the next element in the disk
+
+All of this can be implemented using a tail-recursive function:
+```scala
+def compact(disk: Disk): Disk =
+  @tailrec
+  def compactRec(disk: Disk, acc: Disk): Disk =
+    if disk.isEmpty then
+      acc
+    else
+      disk.head match
+        case None => compactRec(disk.last +: disk.tail.init, acc) // Take the last element, put it first and eliminate free block
+        case file@Some(_) => compactRec(disk.tail, acc :+ file) // Append the file block
+  compactRec(disk, Vector.empty)
+```
+
+### Part 2
+The code remains very similar to `part1`. However this time, the `Disk` structure can't consider characters individually anymore. Consecutive file blocks are indivisible, they form a single `Block`. Thus, we define a new `Block` enumeration. All `Block`s have a size, but `Free` blocks do not have any index attached whereas `File` blocks do:
+```scala
+enum Block(val size: Int):
```

Comment:
Just a cosmetic issue, but here you use 4 spaces for an indent while everywhere else it's 2.